sdk/ai/Azure.AI.Inference/src/Telemetry/OpenTelemetryScope.cs

```diff
@@ -0,0 +1,247 @@
+            SetTagMaybe(GenAiRequestTemperatureKey, requestOptions.Temperature);
+            if (requestOptions.AdditionalProperties != null && requestOptions.AdditionalProperties.ContainsKey("top_p"))
+            {
+                SetTagMaybe(
+                    GenAiRequestTopPKey,
+                    JsonSerializer.Deserialize<double?>(requestOptions.AdditionalProperties["top_p"]));
+            }
+            // Log all messages as the events.
+            foreach (ChatRequestMessage message in requestOptions.Messages)
+            {
+                TagList requestTags = new() {
+                    { GenAiSystemKey, GenAiSystemValue},
+                    { GenAiEventContent, GetContent(message)}
+                };
+                m_activity.AddEvent(
+                    new ActivityEvent(
+                        $"gen_ai.{message.Role}.message",
+                        DateTimeOffset.Now,
+                        new ActivityTagsCollection(requestTags)
+                    )
+                );
+            }
+            m_duration = Stopwatch.StartNew();
+        }
+
+        public void RecordResponse(ChatCompletions response)
+        {
+            RecordResponseInternal(new SingleRecordedResponse(response));
+        }
+
+        public void UpdateStreamResponse<T>(T item)
+        {
+            if (typeof(T) != typeof(StreamingChatCompletionsUpdate))
+                return;
+            StreamingChatCompletionsUpdate castedItem = item as StreamingChatCompletionsUpdate;
+            m_recordedStreamingResponse.Update(castedItem);
+        }
+
+        public void RecordStreamingResponse()
+        {
+            RecordResponseInternal(m_recordedStreamingResponse);
+        }
+
+        /// <summary>
+        /// Log the error.
+        /// </summary>
+        /// <param name="e">Exception thrown by completion call.</param>
+        public void RecordError(Exception e)
+        {
+            if (m_activity == null)
+                return;
+            s_duration.Record(m_duration.Elapsed.TotalSeconds, m_commonTags);
+            var errorType = e?.GetType()?.FullName;
+            if ( e is Azure.RequestFailedException requestFailed)
+            {
+                errorType = requestFailed.Status.ToString();
+            }
+            m_activity.SetTag(ErrorTypeKey, errorType);
+            m_activity.SetStatus(ActivityStatusCode.Error, e?.Message ?? errorType);
+        }
+
+        /// <summary>
+        /// Record the events and metrics associated with the response.
+        /// </summary>
+        /// <param name="recordedResponse"></param>
+        private void RecordResponseInternal(AbstractRecordedResponse recordedResponse)
+        {
+            if (m_activity == null)
+                return;
+            TagList tags = m_commonTags;
+            // Find index of model tag.
+            object objOldModel = m_activity.GetTagItem(GenAiResponseModelKey);
```

Comment:
The response model is the same on all chunks, let's not spend CPU cycles on overwriting it (i.e. let's organize code in the way that populates properties on the activity once when response/stream is received to the end (or is ending with response).

Reply:
This code is executed once, when all chunks were assembled into the Recorded response. The problem is that the Model set on the request may not be the same as the one on the response. For example, `got-4o` mat be converted to `got-4o-2024-05-01-preview`